# 1 - Source: 

Github repo: https://github.com/tuanpham96/DeepLearningSystems-Fall2021/tree/main/HW3

Colab link: https://colab.research.google.com/drive/1eNqRC7M8SzoLPqAcFp9ptWqUmlRjNyuY?usp=sharing

Experiment preparations and runs in these: https://github.com/tuanpham96/DeepLearningSystems-Fall2021/tree/main/HW3/notebooks

# 2 - Process data 

## 2.0. Startup

In [1]:
%%capture
!pip install googletrans==4.0.0rc1 

In [2]:
%%capture

from google.colab import drive
drive.mount("/content/drive")

%cd "/content/drive/MyDrive/Courses/Fall 2021/dlsys/DeepLearningSystems-Fall2021/HW3"

In [3]:
import os
from glob import glob
import pprint
import numpy as np 
import pandas as pd 

from src.display_results import *
from googletrans import Translator

In [4]:
OUTPUT_PATH = 'output'
google_translator = Translator()

## 2.1. Experiments with EN-ES transformers

In [5]:
#@title Load and process data
output_files = glob(OUTPUT_PATH +'/transformer-*.pkl')
dest_lang = 'es' 
es_main_title = 'Experiments with EN-ES transformers'
cols2drop = ['FFN_units','dropout_rate']
es_df, es_final_df, es_base_fdf, es_all_translations = obtain_result(output_files, google_translator, dest_lang, cols2drop)

In [6]:
#@title Losses and accuracies during training
#@markdown click on each button for each experiment/variation
plot_train_progress(es_df, es_final_df, es_base_fdf, es_main_title)

In [7]:
#@title Parameters defining **EN-ES Baseline** model
es_base_fdf.reset_index(drop=True)

exp_name param_name param_val  ... total_time  accuracies    losses
0  baseline       None      None  ...   0.997038   37.268609  0.702587

[1 rows x 15 columns]

In [8]:
#@title Results for all experiments at the *end* of training for **EN-ES**
#@markdown Accuracies are only subtracted normalized, while losses and total training time are percentage normalized to baseline
#@markdown ```
#@markdown accuracies[norm]         = accuracies[experiment] - accuracies[baseline]
#@markdown losses[norm_perc]        = 100*(losses[experiment] - losses[baseline])/losses[baseline]
#@markdown total_time[norm_perc]    = 100*(total_time[experiment] - total_time[baseline])/total_time[baseline]
#@markdown ```

es_cols2drop = ['model_id', 'd_model', 'n_layers', 'n_heads', 'act_fun', 'include_pos_enc', 'vocab_size_enc', 'vocab_size_dec']
es_final_df.drop(es_cols2drop, axis=1).reset_index(drop=True)

exp_name         param_name  ... losses[norm_perc] total_time[norm_perc]
0        baseline               None  ...          0.000000              0.000000
1      vary_depth           n_layers  ...        -26.481614            -37.708414
2      vary_depth           n_layers  ...         25.912865             40.562326
3      vary_depth           n_layers  ...         67.390335             83.915256
4   vary_embeddim            d_model  ...          2.169019              2.949532
5   vary_embeddim            d_model  ...          0.929955              1.286821
6   vary_embeddim            d_model  ...         13.211780              4.937030
7   vary_numheads            n_heads  ...          5.025184              2.678420
8   vary_numheads            n_heads  ...         -0.784877              1.758950
9   vary_numheads            n_heads  ...         -3.195526              2.421746
10     half_vocab  vocab_size_factor  ...        -15.141969             -3.871558
11  remove_posenc    include_pos_enc  ...         -0.509313              4.000035
12    vary_actfun            act_fun  ...         -1.294207              3.782919
13    vary_actfun            act_fun  ...         -3.338568              4.300508

[14 rows x 10 columns]

## 2.2. English-Vietnamses transformer

In [9]:
#@title Load and process data
output_files = glob(OUTPUT_PATH +'/engviet-transformer-*.pkl')
dest_lang = 'vi' 
vi_main_title = 'Experiments with EN-VN transformers'

vi_df, vi_final_df, vi_base_fdf, vi_all_translations = obtain_result(output_files, google_translator, dest_lang)

In [10]:
#@title Losses and accuracies during training
#@markdown click on each button for each experiment/variation
plot_train_progress(vi_df, vi_final_df, vi_base_fdf, vi_main_title)

In [11]:
#@title Parameters defining **EN-VN Baseline** model
vi_base_fdf.reset_index(drop=True)

desc  exp_name param_name  ... total_time accuracies    losses
0  engviet-transformer  baseline       None  ...     0.9533  66.874409  0.173871

[1 rows x 18 columns]

In [12]:
#@title Results for all experiments at the *end* of training for **EN-VN**

vi_cols2drop = ['model_id', 'd_model', 'n_layers', 'n_heads', 'act_fun', 'include_pos_enc', 'vocab_size_enc', 'vocab_size_dec', 'FFN_units', 'dropout_rate']
vi_final_df.drop(vi_cols2drop, axis=1).reset_index(drop=True)

desc       exp_name  ... losses[norm_perc] total_time[norm_perc]
0  engviet-transformer       baseline  ...          0.000000              0.000000
1  engviet-transformer  vary_ffnunits  ...         11.653451              3.597891
2  engviet-transformer  vary_ffnunits  ...          6.061190              3.138718
3  engviet-transformer   vary_dropout  ...        -25.223008              0.506972
4  engviet-transformer   vary_dropout  ...        305.659846              3.230167

[5 rows x 11 columns]

# 3 - Write up
Read the [instruction](https://docs.google.com/document/d/1TeRDGWtfSYa08N30R4jJLboLub11WYzT/edit?usp=sharing&ouid=102957549596160300860&rtpof=true&sd=true) carefully. Document your findings in the following sections.

## 3.0. Note

Due to Colab limited run time and needing attending to, I resorted to training the models using Kaggle instead. 

And apparently, they were usually faster than Colab.

For comparison with translations, since I do not know Spanish, I fed the output of the transformers' predictions to Google Translate back to English and compare with the original input sentences. 

## 3.1. Run the baseline model, understand what it's doing and make sure it works


- Baseline Accuracy: **37.27%** (at the end)
- Baseline Loss: **0.7026** (at the end)
- How long did it take to train? **0.997 hrs** (only measured for `main_train`)

These results are comparable to the tutorial notebook (also 10 epochs), except for the training time, which is faster on Kaggle (roughly 6 min/epoch) than on Colab (roughly 8.5-9 min/epoch). 

As for the translations, it seems that the baseline transformer could only handle short simple sentences (0-3). If the sentences have specialized language (4, 5) or long (6, 7), the transformer only outputed much smaller sentences, and sometimes they don't make sense.

In [13]:
#@title Baseline model
es_base_fdf.reset_index(drop=True)

exp_name param_name param_val  ... total_time  accuracies    losses
0  baseline       None      None  ...   0.997038   37.268609  0.702587

[1 rows x 15 columns]

In [14]:
#@title Baseline translation
print(es_all_translations['transformer-00']['result_text'])

(0) INPUT: 		you should pay for it.
    INPUT  --> TRANSF: 	Deberías pagar por ello.
    INPUT  --> GOOGLE: 	Debes pagar por ello.
    TRANSF --> GOOGLE: 	You should pay for it.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(1) INPUT: 		we have no extra money.
    INPUT  --> TRANSF: 	No tenemos dinero extra.
    INPUT  --> GOOGLE: 	No tenemos dinero extra.
    TRANSF --> GOOGLE: 	We do not have extra money.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(2) INPUT: 		This is a problem to deal with.
    INPUT  --> TRANSF: 	Este problema es un problema de que lo que al problema.
    INPUT  --> GOOGLE: 	Este es un problema para tratar.
    TRANSF --> GOOGLE: 	This problem is a problem t

## 3.2. Experiment with changing the following:

For each experiment where you vary hyperparameters use **at least three different values**.

Explain the results you get **in terms of your understanding of transformers**.


### a) Vary the depth of the network and compare results to the baseline

Here I interpret *depth* as `n_layer` variations: [2, 4, 6, 8] where baseline model has 4 layers. To my surprise, more layers actually **degrade** performance (increase loss by 20-30% and decrease accuracy by around 3% per every 2 layers increase), and of course also increase training time. I really thought that increasing depth would allow the model to learn more effectively. Maybe the models need to be trained for longer and increase `dropout_rate` (though the original paper also has `dropout_rate=0.1`) or apply further regularization to avoid overfitting. One thing I'm not sure of is whether the label smoothing step mentioned in the paper is actually implemented here. That might make a difference.

Additionally, from the translation below for (input-2) and the baseline translation of (input-3), I wonder if there's some inherent gender bias in the model. 

In [15]:
#@title Results
display_df = es_final_df.query('exp_name == "baseline" | exp_name == "vary_depth"')\
    .drop(es_cols2drop, axis=1).reset_index(drop=True)
prm_name = display_df.param_name.tolist()[-1]
display_df.loc[display_df.exp_name == 'baseline','param_name'] = prm_name
display_df.loc[display_df.exp_name == 'baseline','param_val'] = es_base_fdf[prm_name].tolist()[0]
display_df.sort_values(by='param_val').reset_index(drop=True)

exp_name param_name  ... losses[norm_perc] total_time[norm_perc]
0  vary_depth   n_layers  ...        -26.481614            -37.708414
1    baseline   n_layers  ...          0.000000              0.000000
2  vary_depth   n_layers  ...         25.912865             40.562326
3  vary_depth   n_layers  ...         67.390335             83.915256

[4 rows x 10 columns]

In [16]:
#@title Best model translation: `n_layers=2`
print(es_all_translations['transformer-01']['result_text'])

(0) INPUT: 		you should pay for it.
    INPUT  --> TRANSF: 	Deberíais pagarlo.
    INPUT  --> GOOGLE: 	Debes pagar por ello.
    TRANSF --> GOOGLE: 	You should pay it.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(1) INPUT: 		we have no extra money.
    INPUT  --> TRANSF: 	No tenemos dinero extra.
    INPUT  --> GOOGLE: 	No tenemos dinero extra.
    TRANSF --> GOOGLE: 	We do not have extra money.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(2) INPUT: 		This is a problem to deal with.
    INPUT  --> TRANSF: 	Este es un problema con él.
    INPUT  --> GOOGLE: 	Este es un problema para tratar.
    TRANSF --> GOOGLE: 	This is a problem with him.
-------------------------------------

### b) Vary the model dimension (embedding vector length) and compare it to the baseline

The embedding dimension `d_model` is varied [128, 256, 512, 1024] where 512 is the value of the baseline model. The table below shows only the final results, which indicate that increased embedding dimension do not improve performance and excessive increase (compare baseline and `d_model=1024`) could end up harming performance significantly, while training time only differ by a few percents. However, what is not shown in this table is that, during the first 3-4 epochs, higher `d_model` actually benefits performance quite clearly (for both loss and accuracy). This again my point towards the need for higher regularization. 


In [17]:
#@title Results
display_df = es_final_df.query('exp_name == "baseline" | exp_name == "vary_embeddim"')\
    .drop(es_cols2drop, axis=1).reset_index(drop=True)
prm_name = display_df.param_name.tolist()[-1]
display_df.loc[display_df.exp_name == 'baseline','param_name'] = prm_name
display_df.loc[display_df.exp_name == 'baseline','param_val'] = es_base_fdf[prm_name].tolist()[0]
display_df.sort_values(by='param_val').reset_index(drop=True)

exp_name param_name  ... losses[norm_perc] total_time[norm_perc]
0  vary_embeddim    d_model  ...          2.169019              2.949532
1  vary_embeddim    d_model  ...          0.929955              1.286821
2       baseline    d_model  ...          0.000000              0.000000
3  vary_embeddim    d_model  ...         13.211780              4.937030

[4 rows x 10 columns]

### c) Vary the number of attention heads and compare to baseline

The number of heads are varies between [4, 8, 16, 32] where baseline model has 8 heads. As I understand, number of heads would allow the model to observe more variations of relations between tokens/words and be better in terms of performance, as well as faster training time. However, the accuracies here only increase very slightly when we have too many heads, and the training time between these different variations also do not vary much. 

In [18]:
#@title Results
display_df = es_final_df.query('exp_name == "baseline" | exp_name == "vary_numheads"')\
    .drop(es_cols2drop, axis=1).reset_index(drop=True)
prm_name = display_df.param_name.tolist()[-1]
display_df.loc[display_df.exp_name == 'baseline','param_name'] = prm_name
display_df.loc[display_df.exp_name == 'baseline','param_val'] = es_base_fdf[prm_name].tolist()[0]
display_df.sort_values(by='param_val').reset_index(drop=True)

exp_name param_name  ... losses[norm_perc] total_time[norm_perc]
0  vary_numheads    n_heads  ...          5.025184              2.678420
1       baseline    n_heads  ...          0.000000              0.000000
2  vary_numheads    n_heads  ...         -0.784877              1.758950
3  vary_numheads    n_heads  ...         -3.195526              2.421746

[4 rows x 10 columns]

### d) Reduce the vocabulary in half and compare it to the baseline

The maximum vocab size for the baseline model is kept fixed at 2^14, then scaled by half for this experiment. This should have decreased performance but actually decreased loss by 15% and increased accuracy by 2.55%. I think the data might have been quite limited and less variable to begin with. 

In [19]:
#@title Results
display_df = es_final_df.query('exp_name == "baseline" | exp_name == "half_vocab"')\
    .drop(es_cols2drop, axis=1).reset_index(drop=True)
prm_name = display_df.param_name.tolist()[-1]
	# vocab_size_factor	
display_df.loc[display_df.exp_name == 'baseline','param_name'] = prm_name
display_df.loc[display_df.exp_name == 'baseline','param_val'] = 1
display_df

exp_name         param_name  ... losses[norm_perc] total_time[norm_perc]
0    baseline  vocab_size_factor  ...          0.000000              0.000000
1  half_vocab  vocab_size_factor  ...        -15.141969             -3.871558

[2 rows x 10 columns]

### e) Remove the positional encoding and compare it to the baseline

The inclusion of positional encoding also does not indicate any performance improvement and cause a slight decrease 0.05% decrease in accuracy, which could have been just noise. Again, this is odd since leaving out positional information should have degraded performance and sequence generation.  

In [20]:
#@title Results
cols2drop = ['model_id', 'd_model', 'n_layers', 'n_heads', 'act_fun', 'include_pos_enc', 'vocab_size_enc', 'vocab_size_dec']
display_df = es_final_df.query('exp_name == "baseline" | exp_name == "remove_posenc"')\
    .drop(es_cols2drop, axis=1).reset_index(drop=True)
prm_name = display_df.param_name.tolist()[-1]
display_df.loc[display_df.exp_name == 'baseline','param_name'] = prm_name
display_df.loc[display_df.exp_name == 'baseline','param_val'] = es_base_fdf[prm_name].tolist()[0]
display_df

exp_name       param_name  ... losses[norm_perc] total_time[norm_perc]
0       baseline  include_pos_enc  ...          0.000000              0.000000
1  remove_posenc  include_pos_enc  ...         -0.509313              4.000035

[2 rows x 10 columns]

## 3.3. compare ReLU against GELU & Swish
Compare the baseline trained with ReLU to that using GELU as described in this paper ([arXiv:1606.08415v4](https://arxiv.org/pdf/1606.08415v4.pdf)) & Swish described in this paper ([arXiv:1710.05941v1](https://arxiv.org/pdf/1710.05941v1.pdf)). Note that GELU (Gaussian Error Linear Units) are what GPT-2 and GPT-J are using instead of ReLU.   See you get better convergence or better loss during training and if any of the translations in your tests have changed. Both GELU and Swish are provided by Tensorflow.

---

The results below show that these activation functions do enhance performance, `gelu` gaining about 0.12% compared to `relu` and `swish` about 0.34% in accuracy. The losses also decrease, by 1.29% for `gelu` and 3.34% for `swish` compared to `relu`. The training time differences are not that different, around 1 hour for all three. 

As for translations, for (input-0,1), `gelu` and `relu` are similar, but `gelu` seems to make more sense than `relu` for (input-2) based on the Google translation of the model's output. (input-3) somehow gets turned to a question by `gelu`. (input-4) differs quite a lot with `gelu` generating something that's not really related, while `relu` still retains the word "interesting". `gelu` at least retains the word "learning" for (5). The (input-6,7) are too hard for both but `gelu` translations, at least via the Google translated version of the model, make more sense than `relu`.

For `swish`, it is interesting that (input-0) is much shorter. (input-1) is similar to that for `relu`.  (input-2) also differs but matches the first 4 words when compared with the Google translation. (input-3,4,6,7) are basically garbage translation, but (input-5) interesting matches really well, missing only 1-2 words. 

In [21]:
#@title Results
display_df = es_final_df.query('exp_name == "baseline" | exp_name == "vary_actfun"')\
    .drop(es_cols2drop, axis=1).reset_index(drop=True)
prm_name = display_df.param_name.tolist()[-1]
display_df.loc[display_df.exp_name == 'baseline','param_name'] = prm_name
display_df.loc[display_df.exp_name == 'baseline','param_val'] = es_base_fdf[prm_name].tolist()[0]
display_df

exp_name param_name  ... losses[norm_perc] total_time[norm_perc]
0     baseline    act_fun  ...          0.000000              0.000000
1  vary_actfun    act_fun  ...         -1.294207              3.782919
2  vary_actfun    act_fun  ...         -3.338568              4.300508

[3 rows x 10 columns]

In [22]:
#@title Baseline translation
print(es_all_translations['transformer-00']['result_text'])

(0) INPUT: 		you should pay for it.
    INPUT  --> TRANSF: 	Deberías pagar por ello.
    INPUT  --> GOOGLE: 	Debes pagar por ello.
    TRANSF --> GOOGLE: 	You should pay for it.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(1) INPUT: 		we have no extra money.
    INPUT  --> TRANSF: 	No tenemos dinero extra.
    INPUT  --> GOOGLE: 	No tenemos dinero extra.
    TRANSF --> GOOGLE: 	We do not have extra money.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(2) INPUT: 		This is a problem to deal with.
    INPUT  --> TRANSF: 	Este problema es un problema de que lo que al problema.
    INPUT  --> GOOGLE: 	Este es un problema para tratar.
    TRANSF --> GOOGLE: 	This problem is a problem t

In [23]:
#@title GELU translation
print(es_all_translations['transformer-12']['result_text'])

(0) INPUT: 		you should pay for it.
    INPUT  --> TRANSF: 	Deberías pagar por eso.
    INPUT  --> GOOGLE: 	Debes pagar por ello.
    TRANSF --> GOOGLE: 	You should pay for that.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(1) INPUT: 		we have no extra money.
    INPUT  --> TRANSF: 	No tenemos dinero extra.
    INPUT  --> GOOGLE: 	No tenemos dinero extra.
    TRANSF --> GOOGLE: 	We do not have extra money.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(2) INPUT: 		This is a problem to deal with.
    INPUT  --> TRANSF: 	Este problema es un problema con mucho de hacer.
    INPUT  --> GOOGLE: 	Este es un problema para tratar.
    TRANSF --> GOOGLE: 	This problem is a problem with a 

In [24]:
#@title Swish translation
print(es_all_translations['transformer-13']['result_text'])

(0) INPUT: 		you should pay for it.
    INPUT  --> TRANSF: 	Deberías pagar.
    INPUT  --> GOOGLE: 	Debes pagar por ello.
    TRANSF --> GOOGLE: 	You should pay.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(1) INPUT: 		we have no extra money.
    INPUT  --> TRANSF: 	No tenemos dinero extra.
    INPUT  --> GOOGLE: 	No tenemos dinero extra.
    TRANSF --> GOOGLE: 	We do not have extra money.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(2) INPUT: 		This is a problem to deal with.
    INPUT  --> TRANSF: 	Este es un problema con nosotros.
    INPUT  --> GOOGLE: 	Este es un problema para tratar.
    TRANSF --> GOOGLE: 	This is a problem with us.
--------------------------------------

## 3.4. English --> Vietnamese

Since this is my language, I at least could have a better sense to evaluate the translation. For the dataset, I also downloaded from the same link, though the dataset is quite small compared to the Spanish version (see below). So I reduced the batch size from 64 to 32 instead, but kept the vocabulary the same. Additionally, since nonbreaking prefixes aren't really a thing in Vietnamese (except for certain usage of abbreviations), or at least not very common, I created a `nonbreaking_prefixes.vn` file with just consonants and vowels (with a few exceptions), as that's what I see common in those files in other languages.  

From the previous experiments with EN-ES, I also adjusted some model parameters (see below for details): reduce `d_model`, increase `n_heads` and change activation function to `swish`. Since the dataset is small, I trained the model for 50 epochs (logged every 25 batches), taking around an hour per model. 

I also played around with `FFN_units` and `dropout_rate`. While the former doesn't seem to have any noticeable effect, increasing the latter, interestingly, decrease performance. I don't have an explanation for this because I thought increasing the rate would regularize the model better. It is possible that the model has to be sufficiently big enough for high dropout rate to be beneficial. 

Regardless, the reported numbers here are from the best model out of the 5 (i.e. `dropout_rate=0.05`). 

- Activation function: **`Swish`**
- Accuracy: **68.15%**
- Loss:	**0.1300**
- How long did it take to train? **0.958 hr**

For the translation results, here's what I got (if I don't mention it, it means that the Google's translations of the original inputs are correct)

- (0) Wrong translation but this is somewhat understandable since "trả" appears in both the Vietnamese words for "pay" and "answer".
- (1) This is basically wrong without any excuse. 
- (2) This is actually a good translation, except not handling the phrase "to deal with"
- (3) This is wrong but the sentiment is roughly there.
- (4, 5) Only the first 2 words match the content, the rest is garbage. 
- (6, 7, 8, 10, 11) The level of wrong is unbelievable. 
- (7) At least the mentioning of the word meaning "lake" is interesting but the grammar and the meaning are basically garbage. 
- (8) Interesting the word meaning "rain" appears but still garbage.



In [25]:
#@title Datasize comparison
#@markdown  a simple `wc -l` for the data files reveals that the Vietnamese version is around 17 smaller than the Spanish version
!wc -l data/eng_spa_translations/spa.txt
!wc -l data/eng_vie_translations/vie.txt

134736 data/eng_spa_translations/spa.txt
7558 data/eng_vie_translations/vie.txt


In [26]:
#@title Parameters defining **EN-VN Baseline** model
vi_base_fdf.reset_index(drop=True)


desc  exp_name param_name  ... total_time accuracies    losses
0  engviet-transformer  baseline       None  ...     0.9533  66.874409  0.173871

[1 rows x 18 columns]

In [27]:
#@title Parameters defining **EN-ES Baseline** model
es_base_fdf.reset_index(drop=True)


exp_name param_name param_val  ... total_time  accuracies    losses
0  baseline       None      None  ...   0.997038   37.268609  0.702587

[1 rows x 15 columns]

In [28]:
#@title Results for all experiments at the *end* of training for **EN-VN**

vi_cols2drop = ['model_id', 'd_model', 'n_layers', 'n_heads', 'act_fun', 'include_pos_enc', 'vocab_size_enc', 'vocab_size_dec', 'FFN_units', 'dropout_rate']
vi_final_df.drop(vi_cols2drop, axis=1).reset_index(drop=True)

desc       exp_name  ... losses[norm_perc] total_time[norm_perc]
0  engviet-transformer       baseline  ...          0.000000              0.000000
1  engviet-transformer  vary_ffnunits  ...         11.653451              3.597891
2  engviet-transformer  vary_ffnunits  ...          6.061190              3.138718
3  engviet-transformer   vary_dropout  ...        -25.223008              0.506972
4  engviet-transformer   vary_dropout  ...        305.659846              3.230167

[5 rows x 11 columns]

In [29]:
#@title Best translation `engviet-transformer-03`
print(vi_all_translations['engviet-transformer-03']['result_text'])

(0) INPUT: 		you should pay for it.
    INPUT  --> TRANSF: 	Bạn nên trả lời điều đó.
    INPUT  --> GOOGLE: 	Bạn nên trả tiền cho nó.
    TRANSF --> GOOGLE: 	You should answer that.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(1) INPUT: 		we have no extra money.
    INPUT  --> TRANSF: 	Chúng ta không có ngốc hơn trước nữa.
    INPUT  --> GOOGLE: 	Chúng tôi không có thêm tiền.
    TRANSF --> GOOGLE: 	We are not stupid than before.
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
(2) INPUT: 		This is a problem to deal with.
    INPUT  --> TRANSF: 	Đây là vấn đề quan trọng
    INPUT  --> GOOGLE: 	Đây là một vấn đề để đối phó với.
    TRANSF --> GOOGLE: 	This is an important issue
-----

## 3.5. Summary

Training transformer is hard. Not based on my understanding but based on the results with the experiments shown here:
- Increased model depth or dropout rate has the most noticeable degrading effect on the model's performance. 
- Increased embedding dimension only benefits initially for training but eventually decreases performance.
- Increased number of heads or changing the simple `relu` to `gelu` or `swish` only results in really tiny boost in performance and doesn't seem to affect the training time. 
- Varying the number units in FFN layer or inclusion of the positional encoding layer does not seem to have any noticeable effect. 
- Having a smaller vocabulary size interestingly gains noticable performance gain. 